# This is an intectual property of the Glowworm Academy

# Any unauthorized publishing,sharing, replicating is subjected to legal prosecustion 

# ------------------------------------------------------------------------------------------------------


## Learning objective of this project

### 1. We will perform a basic level of data analysis on the data.
### 2. finding out about each variable their types and various basic statistical parameters about the data.
### 3. Using the pandas library we will be learning how to manipulate the dataframe by extracting the subset of the data, findin the missing values, finding unique values in thedata
### 4. Drawing conclusions.

In [6]:
#Importing libraries

import pandas as pd
import matplotlib.pyplot as plt


In [7]:
#Importing the database
df = pd.read_csv("insurance.csv")

# This dataset indicates the insurance charges based on the candidate's gender, number of childre, age, smoking habits , region and BMI.

# Task 1: Let's do some prelimnary analysis on the dataset.

In [8]:
#Let's look at the first five rows of the dataset.

df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [9]:
#Let's look at the last five rows of the dataset.

df.tail()

age     sex    bmi  children smoker     region     charges
1333   50    male  30.97         3     no  northwest  10600.5483
1334   18  female  31.92         0     no  northeast   2205.9808
1335   18  female  36.85         0     no  southeast   1629.8335
1336   21  female  25.80         0     no  southwest   2007.9450
1337   61  female  29.07         0    yes  northwest  29141.3603

In [10]:
#Let's find out how many number of rows the dataset has

print("The dataset has the number of rows: ",df.shape[0])

The dataset has the number of rows:  1338


In [11]:
#Let's find out how many number of columns the dataset has

print("The dataset has the number of columns: ",df.shape[1])

The dataset has the number of columns:  7


In [12]:
#Let's find out the column titles of the dataset

print("Following are the column names of for the dataset: ")

print(df.columns)

Following are the column names of for the dataset: 
Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')


In [13]:
#To print in a better way we will use the for loop

for i in df.columns:
    print(i)

age
sex
bmi
children
smoker
region
charges


# Explaination of the all the column titles

## age - Age of the the candidate 
## sex - Gender of the the candidate 
## bmi - Body Mass Index of the the candidate 
## children - Number of children the candidate has
## smoker - Whether the candidate smokes or not
## region - Which region the candidate belongs to
## charges - What is the insurance cost of the candidate 



In [15]:
#Let's use the .info method on the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


# Conclusion

## 1. All the columns except the sex, smoker and region has numerical data.
## 2. Total there are 1338 rows.
## 3. There are no empty values in the dataset.

In [16]:
#Let's get the statistical information on the dataset

df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

# Conclusion

## 1. Average age of the candidate is 39 years with a BMI value of 30.
## 2. High standard deviation in the age and BMI indicates that the value of age and BMI is varying significantly in the dataset. 
## 3. The youngest candidate is 18 years old and the oldest is 64 years old. (High variation)
## 4. The candidates BMI range from 15 to 53. (High variation)
## 5. High standard deviation in the charges is varying significantly in the dataset. This means that the current variables may not be enough to predict the insurance charges in the dataset.

# It is important to note that .describe() methods gives important statistical metrics only for columns which have a numerical values.

# Task 2: Let's learn how to extract a subset of the dataset

### For example out of we may need to look at only few columns to look at the data more specifically.

In [17]:
#Printing only the age, sex and the bmi columns

df[['age','sex','bmi']]

age     sex     bmi
0      19  female  27.900
1      18    male  33.770
2      28    male  33.000
3      33    male  22.705
4      32    male  28.880
...   ...     ...     ...
1333   50    male  30.970
1334   18  female  31.920
1335   18  female  36.850
1336   21  female  25.800
1337   61  female  29.070

[1338 rows x 3 columns]

In [18]:
#Printing only the charges and smoke columns

df[['charges','smoker']]

charges smoker
0     16884.92400    yes
1      1725.55230     no
2      4449.46200     no
3     21984.47061     no
4      3866.85520     no
...           ...    ...
1333  10600.54830     no
1334   2205.98080     no
1335   1629.83350     no
1336   2007.94500     no
1337  29141.36030    yes

[1338 rows x 2 columns]

In [19]:
#Let's extract rows from 100 to 700 and all the columns

df.iloc[100:701,:]

age     sex     bmi  children smoker     region      charges
100   41  female  31.600         0     no  southwest   6186.12700
101   30    male  25.460         0     no  northeast   3645.08940
102   18  female  30.115         0     no  northeast  21344.84670
103   61  female  29.920         3    yes  southeast  30942.19180
104   34  female  27.500         1     no  southwest   5003.85300
..   ...     ...     ...       ...    ...        ...          ...
696   53  female  32.300         2     no  northeast  29186.48236
697   41    male  35.750         1    yes  southeast  40273.64550
698   56    male  33.725         0     no  northwest  10976.24575
699   23  female  39.270         2     no  southeast   3500.61230
700   21  female  34.870         0     no  southeast   2020.55230

[601 rows x 7 columns]

In [20]:
#Let's extract rows from 100 to 700 for smoker, region and charges

df.iloc[100:701,4:7]

smoker     region      charges
100     no  southwest   6186.12700
101     no  northeast   3645.08940
102     no  northeast  21344.84670
103    yes  southeast  30942.19180
104     no  southwest   5003.85300
..     ...        ...          ...
696     no  northeast  29186.48236
697    yes  southeast  40273.64550
698     no  northwest  10976.24575
699     no  southeast   3500.61230
700     no  southeast   2020.55230

[601 rows x 3 columns]

# Task3: Let's find out how many NaN (Not a Number) are there in the dataset

### 1. We will be defining an empty list.
### 2. We will be running a for loop over each of the columns in the dataset.
### 3. For each of the column we will be using the pandas method .isna() to find out if a particular cell is empty or not.
### 4. For each column, the loop will scan through each cell and find out the total number of times it has encountered an empty cell and sum it.
### 5. Finally, it will give us in the form of a list empty cells for all the columns in the dataset.

In [21]:
#Defining the empty list
lst_nan = []

for i in range(len(df.columns)):
    lst_nan.append(df.iloc[:,i].isna().sum())

In [22]:
#Printing the list
lst_nan

[0, 0, 0, 0, 0, 0, 0]

In [23]:
#Converting that into a dataframe
df_nan = pd.DataFrame(lst_nan,index = df.columns)

In [24]:
df_nan

0
age       0
sex       0
bmi       0
children  0
smoker    0
region    0
charges   0

# Conclusion

## 1. There are no empty values in the dataset. This was noted from the df.info() line however, this approach is typically useful in the case where one needs to identify the null values for each column.

# Task4: Let's find out how many unique values are there in the dataset

### 1. We will be defining an empty list.
### 2. We will be running a for loop over each of the columns in the dataset.
### 3. For each of the column we will be using the pandas method .nunique() to find out how many unique values are there in that column.
### 4. For each column, the loop will scan through each cell and find out the total number of times it has encountered an unique value.
### 5. Finally, it will give us the number of unique values in the form of a list for all the columns in the dataset.

In [33]:
#Defining the empty list
lst_unique = []

for i in range(len(df.columns)):
    lst_unique.append(df.iloc[:,i].nunique())

In [35]:
#Printing the list
lst_unique

[47, 2, 548, 6, 2, 4, 1337]

In [36]:
#Converting that into a dataframe
df_unique = pd.DataFrame(lst_unique,index = df.columns)

In [37]:
df_unique

0
age         47
sex          2
bmi        548
children     6
smoker       2
region       4
charges   1337

# Conclusion

## Unique values help the analyst to determine whether a particular variable (column) should be treated as a continous variable or a categorical variable. 

## Lower unique values mean that using that particular variable the data can be divided into categories and vice versa. Typically true for variables such as gender, region etc.

## 1. Highest number of unique values are in the charges column. This means this can be treated as a continuous variable.
## 2. Lowest number of unique values are in the sex and the smoker column followed by region. This means these can be treated as a categorical variable.
## 3. The number of children can also be treated as a categorical variable.
## 4. The bmi having higher number of unique values can be treated as a continuous variable.